_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+and if we encounter a borrowed string, slice, or opaque, we have to remember
+not only which lifetime it corresponds to, but also where in the type tree it
+appeared. Take the following example:
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Opaque {
+        // ...
+    }
+
+    pub struct Second<'a> {
+        data: &'a Opaque,
+    }
+
+    pub struct First<'a> {
+        second: Second<'a>,
+    }
+
+    impl<'a> First<'a> {
+        fn get_data(first: First<'a>) -> &'a Opaque {
+            first.second.data
+        }
+    }
+}
+```
+Remember that Diplomat only looks at function signatures, so we can't just look
+at the `first.second.data` staring the user in the face. What happens instead
+is that we see that `'a` is in the output type, `&'a Opaque`. Then, we traverse
+the inputs and see `first: First<'a>`. We see that it's a non-opaque struct,
+so we look into its fields and find `second: Second<'a>`. That's also a
+non-opaque struct, so we look into _its_ fields and see `data: &'a Opaque`.
+Since `Opaque` is opaque, Diplomat isn't concerned with its fields, so we're
+done recursing. Once we reach this point, we then have to remember that
+the output borrows from `first.second.data`. In the code generation, we can
+then reflect this information, making the C++ example above not generate
+overly conservative documentation.
+
+This is only half the problem, though. What about non-opaque borrowing structs
+in the output?
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Opaque {
+        // ...
+    }
+
+    pub struct Input<'a> {
+        data: &'a Opaque,
+    }
+
+    pub struct Output<'a> {
+        data: &'a Opaque,
+    }
+
+    impl<'a> Input<'a> {
+        fn get_data(input: Input<'a>) -> Output<'a> {
+            let out = Output { data: input.data }
+            out
+        }
+    }
+}
+```
+This follows similarly to the previous example, except we observe that
+`out.data` has to live `'a`, and that `input.data` is guaranteed to live `'a`,
+so we can just enforce that `input.data` has to outlive `out.data` to uphold
+the invariant that `out.data` lives at least `'a`.
+
+## Challenge 3: Index-based Lifetimes
+Tracking the "outlives" relationship on a field-by-field basis is great, but
+it introduces another challenge: lifetimes can go by different names in
+different declarations:
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Opaque {
+        // ...
+    }
+
+    pub struct Input<'i> {
+        data: &'i Opaque,
+    }
+
+    pub struct Output<'o> {
+        data: &'o Opaque,
+    }
+
+    impl<'a> Input<'a> {
+        fn get_data(input: Input<'a>) -> Output<'a> {
+            let out = Output { data: input.data }
+            out
+        }
+    }
+}
+```
+The lifetimes in the declarations of `Input` and `Output` have completely
+different names. We can't just track `'a` all the way down, which sucks because
```

Comment:
suggestion: Probably say something like "We can't just continue looking for lifetimes named `'a` all the way down since `'a` is being mapped to other lifetimes. Instead, we track lifetimes by the index they appear in the arguments, and use that to figure out which lifetime is which. This is similar to what rustc does."

(you don't _have_ to track lifetimes by index, it's just easier here.)